spec/middleware-1.0.0-draft.1.md

```diff
@@ -0,0 +1,309 @@
+application frameworks provide such abstractions in the form of reusable
+software components.
+
+This specification provides a standardised signature for such components in the
+OWIN ecosystem, hereby referred to as **Middleware**.
+
+OWIN Middleware is defined in terms of a delegate structure building upon the
+OWIN: Open Web Server Interface for .NET [Owin 1.0], with no
+dependency on external libraries.
+
+### 1.1. Requirements Notation
+
+ The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC 2119][rfc2119].
+
+A component is deemed to be compliant with this specification if all absolute
+requirements and prohibitions are conformed to.
+
+Components and terms defined in this and other specifications are always
+capitalized, so as to clarify between a general term and the term as used in a
+specification.
+
+## 2. Definitions
+
+This document refers to the following terms:
+
+ - **Application**: a process or operation inspecting, modifying or otherwise
+   accessing HTTP messages, as an AppFunc signature.
+
+ - **Middleware**: Pass through components, capable of mediating request and
+   response messages, in a pipeline model which, when
+   composed, forms an Application.
+
+ - **Builder**: The extension point lambda used by developers to add Middleware
+   components to a pipeline.
+
+ - **Server**: As per the OWIN specification, an HTTP server that directly
+   communicates with the client, and uses OWIN semantics to process requests.
+
+## 3. Middleware
+
+An OWIN Application (as defined by an AppFunc) can be modelled as
+Middleware components layered in a pipeline.
+
+Each Middleware receiving a request may perform some processing on the request,
+may forward the request to the next Middleware and return the resulting
+response, or may decide to return a response immediately and not call the rest
+of the chain.
+
+In other words, a Middleware functions as a proxy of sort, completely
+encapsulating and hiding the next Middleware in the chain, transparently to the
+caller.
+
+### 3.1. Examples
+
+For example, a Middleware responsible for logging all requests and responses
+could execute as the first component of a pipeline, and log all requests and
+responses as they pass through.
+
+```
+        request      > (log request)      > (process request)
+SERVER =============== LOGGER ============= WEB APPLICATION
+        response     < (log response)     < (return response)
+```
+
+Other types of Middleware would respond immediately, without calling the next
+component in the pipeline, such as for an authorization module denying a request
+before a web application can respond.
+
+```
+        request      > (authenticate)
+SERVER =============== LOGGER ------X------ WEB APPLICATION
+        response     < (deny)
+```
+
+### 3.2. Middleware Operation
```

Comment:
If we add those, we need to fine what "producing a response" means in the context of the owin spec. 
I'd suggest for request fields that a CAN would be better than a MAY.
The response header / stream is about modifications done post "producing a response", whatever definition we go with. The specs do not define what happens when a response has been  produced (aka headers sent), does not preclude writing to the response stream after other middlewares are finished but before the stream is closed (chunked encoding), and some implementations may well use trailing headers, in which case writing response headers is perfectly reasonable after closing a stream. I'd personally prefer any information to not have normative value, and be maybe an implementation note, with "Writing to response headers or the response entity body stream after a response has been produced is undefined by this specification".
The owin.ResponseHeaders being one instance only and not to be modified for the server to understand it is mighty fine, but that belongs in the owin spec, not the middleware one. Any app could change those keys in that spec.